src/analysis/monotone-analyzer-impl.h

```diff
@@ -71,76 +60,72 @@ template<size_t N> inline void BlockState<N>::print(std::ostream& os) {
   os << std::endl;
   auto cfgIter = cfgBlock->rbegin();
 
+  // Since we don't store the intermediate states in the BlockState, we need to
+  // re-run the transfer function on all the CFG node expressions to reconstruct
+  // the intermediate states here.
   while (cfgIter != cfgBlock->rend()) {
-    // run transfer function.
     os << ShallowExpression{*cfgIter} << std::endl;
-    BlockState<N>::visit(*cfgIter);
+    BlockState::visit(*cfgIter);
     currState.print(os);
     os << std::endl;
     ++cfgIter;
   }
 }
 
-template<size_t N>
-MonotoneCFGAnalyzer<N> inline MonotoneCFGAnalyzer<N>::fromCFG(CFG* cfg) {
-  MonotoneCFGAnalyzer<N> result;
-
-  // Map BasicBlocks to each BlockState's index
+template<typename Lattice>
+inline void MonotoneCFGAnalyzer<Lattice>::fromCFG(CFG* cfg) {
+  // Construct BlockStates for each BasicBlock and map each BasicBlock to each
+  // BlockState's index in stateBlocks.
   std::unordered_map<const BasicBlock*, size_t> basicBlockToState;
   size_t index = 0;
   for (auto it = cfg->begin(); it != cfg->end(); it++) {
-    result.stateBlocks.emplace_back(&(*it));
+    stateBlocks.emplace_back(&(*it), lattice);
     basicBlockToState[&(*it)] = index++;
   }
 
   // Update predecessors and successors of each BlockState object
   // according to the BasicBlock's predecessors and successors.
-  for (index = 0; index < result.stateBlocks.size(); ++index) {
-    BlockState<N>& currBlock = result.stateBlocks.at(index);
+  for (index = 0; index < stateBlocks.size(); ++index) {
+    BlockState<Lattice>& currBlock = stateBlocks.at(index);
     BasicBlock::Predecessors preds = currBlock.cfgBlock->preds();
     BasicBlock::Successors succs = currBlock.cfgBlock->succs();
     for (auto pred : preds) {
-      currBlock.addPredecessor(&result.stateBlocks[basicBlockToState[&pred]]);
+      currBlock.predecessors.push_back(&stateBlocks[basicBlockToState[&pred]]);
     }
 
     for (auto succ : succs) {
-      currBlock.addSuccessor(&result.stateBlocks[basicBlockToState[&succ]]);
+      currBlock.successors.push_back(&stateBlocks[basicBlockToState[&succ]]);
     }
   }
-
-  return result;
 }
 
-template<size_t N> inline void MonotoneCFGAnalyzer<N>::evaluate() {
+template<typename Lattice>
+inline void MonotoneCFGAnalyzer<Lattice>::evaluate() {
   std::queue<Index> worklist;
 
   for (auto it = stateBlocks.rbegin(); it != stateBlocks.rend(); ++it) {
     worklist.push(it->index);
   }
 
   while (!worklist.empty()) {
-    BlockState<N>& currBlockState = stateBlocks[worklist.front()];
+    BlockState<Lattice>& currBlockState = stateBlocks[worklist.front()];
     worklist.pop();
+
+    // Run transfer function on the block.
```

Comment:
This comment doesn't seem too helpful since the method is already called `transfer`. It may be useful to give a higher-level explanation of what this means, though.

Reply:
Updated.